src/Space.cpp

```diff
@@ -348,8 +352,20 @@ void Space::ToJson(Json &jsonObj)
 	spaceObj["frame"] = frameObj;
 
 	Json bodyArray = Json::array(); // Create JSON array to contain body data.
-	for (Body *b : m_bodies) {
+	for (size_t i = 1; i < m_bodyIndex.size(); i++)
+	{
+		Body* b = m_bodyIndex[i];
```

Comment:
At this point, I'm tempted to suggest adding a `bodyID` value directly to the `Body` object itself and store body pointers in an efficient sparse set/pool instead of a linear array. We've needed a unique slot+generation identifier for Bodies for quite some time (the BodyComponent system would benefit significantly) and it seems this PR implements that concept here in a limited fashion already.

Reply:
This will make easier tracking of all objects that's in the game at the moment (even if they're not in space) and will most likely eliminate the need of `m_bodyIndex` and index rebuilding during saving and loading the game.
And to continue, I would actually just leave only `m_bodies` array and added functions like `GetBodiesInSpace` (for `Update`) and `GetAllBodies` (for save/load).
Do you have any suggestions for generating unique id or just simple index (like in this `for` loop) will suffice for now?

Reply:
I'd highly recommend the use of a `{ generation, slot id }` style identifier, rather than a monotonically increasing integer. You'd allocate a `std::vector<Body*> m_bodies` as current, and allocate a `std::vector<uint16_t> m_generation` - all generation indices are initialized to 1. When a body is destroyed, clear the slot in the array and increment the generation index for that slot (so existing references to the body ID are visibly invalid). This can use 32-bit or 64-bit values for the identifier (e.g. 20bit + 12bit slot + generation pair packed into 32 bits).

Implementation is fairly simple, I'll see if I can write some example code for achieving this in the next few days and publish it on a branch on my fork. The upside is this avoids existing erasure overhead when deleting bodies, and makes it very simple to port the BodyComponent implementation to use Body IDs in a sparse set.

Reply:
Although not showed here, I added the `FIXME` when creating empty body object with a brief description why it's done and link to your [comment](https://github.com/pioneerspacesim/pioneer/pull/5657#issuecomment-1818188703).